src/hooks/useSponsoredDelegation.ts

```diff
@@ -0,0 +1,78 @@
+import { useNonce } from "@/hooks/useNonce";
+import Tenant from "@/lib/tenant/tenant";
+import { UIGasRelayConfig } from "@/lib/tenant/tenantUI";
+import { Address } from "viem";
+import AgoraAPI from "@/app/lib/agoraAPI";
+import { useSignTypedData } from "wagmi";
+import { DelegateChunk } from "@/app/api/common/delegates/delegate";
+import { useState } from "react";
+
+interface Props {
+  address: `0x${string}` | undefined;
+  delegate: DelegateChunk;
+}
+
+const types = {
+  Delegation: [
+    { name: "delegatee", type: "address" },
+    { name: "nonce", type: "uint256" },
+    { name: "expiry", type: "uint256" },
+  ],
+};
+
+export const useSponsoredDelegation = ({ address, delegate }: Props) => {
+  const { ui, contracts } = Tenant.current();
+  const { signTypedDataAsync } = useSignTypedData();
+
+  const [isFetching, setIsFetching] = useState(false);
+  const [isFetched, setIsFetched] = useState(false);
+  const [txHash, setTxHash] = useState<`0x${string}` | undefined>(undefined);
+
+  const isGasRelayEnabled = ui.toggle("sponsoredDelegate")?.enabled === true;
+  const gasRelayConfig = ui.toggle("sponsoredDelegate")!
+    .config as UIGasRelayConfig;
+
+  const { data: nonce } = useNonce({
+    address: address,
+    enabled: isGasRelayEnabled && !!address,
+  });
+
+  const call = async () => {
+    setIsFetching(true);
+    setIsFetched(false);
+
+    const latestBlock = await contracts.token.provider.getBlock("latest");
+    const expiry = (latestBlock?.timestamp || 0) + 1000;
+
+    const signature = await signTypedDataAsync({
+      domain: {
+        name: gasRelayConfig.domain,
```

Comment:
I feel like this shouldn't be part of the config -- it's always going to be the name of the token, right? So we can just read the `name` property off of the token contract, then use that? Addresses the comment you left in the config file itself

Reply:
brilliant. I wish I had thought about that 🤦